rfc/project-cutover.md

```diff
@@ -0,0 +1,109 @@
+* Each microservice container (owned by `spinnaker-marketplace`) - all containers registered in Google Container Registry (GCR), with each layer being stored in Google Cloud Storage (GCS)
+* `spinaker-artifacts` bucket (owned by `spinnaker-marketplace`) - all released `spin` binaries, Halyard binaries, and `sponnet` library.
+* `spinnakerbot` GKE cluster (owned by `spinnaker-build`) - Github bot for various issue and pull-request management runs here.
+
+### Goals and Non-Goals
+
+Goals:
+* Move all resulting artifacts to the new project with minimal disruption
+
+Non-Goals:
+* Change *how* the artifacts are built or published - just where they are done. including:
+  * Start using Github Package Repo
+  * Use Google Cloud Soure Repo during container builds
+  * Use of Jenkins to orchestrate the release process
+
+## Motivation and Rationale
+
+_What is the current state of the world?_
+* Docker containers are orchestrated from a Jenkins instance in `spinnaker-community` and actually built, stored, and served from the `spinnaker-marketplace` project.
+  * A nightly BOM and each release BOM is generated and stored in the `halconfig` bucket, along with copies of all microservice config files from their respective `halconfig/` directories in each repo.
+
+* Halyard, spin, and sponnet releases are all orchested from the same Jenkins instance and are stored in their respective locations in `spinnaker-artifacts`
+
+* `spinnakerbot` can be [built and deployed](https://github.com/spinnaker/spinnaker/blob/master/spinbot/Makefile#L11) by any developers with Google Container Builder (GCB) priviledges on the `spinnaker-marketplace` project.
+
+_Why is this change being proposed?_
+
+* Transparency in the process, remove Google as the sole owner/operator/bottleneck of the whole thing.
+
+_Who have you identified as early customers of this change?_
+
+* End-users and any new non-Google engineers that will be involved in the expanded nightly build-cop and release-manager duties.
+
+## Timeline
+
+_What is the proposed timeline for the implementation?_
+
+ASAP
+
+## Design
+
+_What exactly are you doing?_
+
+
+### `halconfig` bucket
+
+GCS buckets exist in a global namespace, and cannot be directly transferred between projects.
+
+Two possible approaches:
+1. (Preferred): Copy all data from bucket `halconfig` in `spinnaker-marketplace` to `halconfig-temp` in `spinnaker-community`. Then delete the old bucket and recreate it in the new project.
+2. Copy all data to `halconfig-v2` in the new project.
+  * Update all Halyard config to use this new bucket
+  * Update build and release config to start publishing here
+  * Update any documentation on spinnaker.io
+
+Approach 1 comes with some downtime as the data is transferred and buckets get renamed. Also comes with very slight risk of someone else grabbing the bucket name inbetween when the old bucket is deleted and the new one with the same name is created.
```

Comment:
+1
Even if the risk is small, it seems like we'd be in a pretty bad situation if we were unable to get that bucket name in the new project.  I'd advocate for approach 2 for this reason, but at the very least would want to see a pretty clear playbook of how we would recover from losing the bucket before even considering approach 1.

Reply:
https://cloud.google.com/storage/docs/moving-buckets describes the process. Move data to temp location, delete bucket, recreate is new project, move data from temp location. 

nit: We wouldn't be "maintaining" the old location - it would just be frozen.

Risks for approach 2 include broken nightly and release process until references to the old bucket location are fixed. We would need to release Halyard with new configs, and bump the minimum Halyard version requirement for releases going forward. 

Also it would break anyone in the community who has built tooling against the current bucket name, as that one would no longer receive updates (not sure if Armory is in this boat). 

IMO, it seems kind of excessive for, if everything went well, probably less than 2 minutes of exposure during the bucket delete and recreate, but I acknowledge that the downside if something does goes wrong would be impactful to the community. Risks of using open source software, I guess?

 I would have a better estimate of full downtime by timing the initial transfer of all files to the temp bucket (there files aren't big, there's just a lot of small files)

Reply:
It does seem unlikely, but a worse case scenario for not getting the bucket name for something like `spinnaker-artifacts` seems like it could create security issues.

Reply:
> probably less than 2 minutes of exposure during the bucket delete and recreate

Can you test this with a bucket that isn't the _real_ bucket? i.e. dump a few hundred GB of data in a `spinaker-rulez-test` bucket and move that across projects and see how it goes?

I feel like approach 1 is way riskier than you're letting on... "it'll probably be fine, I think?" is basically all the assurance this doc offers...

Reply:
All data from halconfig was copied in just over 3 minutes:

```
$ time gsutil -m cp -r gs://halconfig/* gs://testing-halconfig2
...
real	3m38.036s
user	2m54.045s
sys	0m31.103s
```

A move (interspersed cp and deletes) went much slower at :
```
$ time gsutil -m mv gs://testing-halconfig2/** gs://testing-halconfig3/
...
real	161m29.141s
user	9m4.821s
sys	1m36.166s
```
most of which is *way* outdated config files from builds in 2017 and 2018, because the build scripts will delete BOMs and docker container images, but leaves the config files. I'm working on a PR to delete these obsolete configs.

Unfortunately I can't give any more assurance - the only way to transfer a bucket between projects is to delete it in the old project and recreate it in the new project. I checked for any internal tools and none exist and are unlikely to be built because of assumptions made in the billing system or something like that.

Reply:
But can you try it with a different bucket? i.e. delete `testing-halconfig2` and see if you're able to immediately recreate it in a different project? I'm worried the name is going to be on hold for some period of time...

Reply:
Also, wrt the time needed to copy, have you timed it using the transfer feature directly in the cloud console?

Reply:
> I'm worried the name is going to be on hold for some period of time...

I've done this many times (including once today), and have never encountered any delay in being able to reuse the name.

Reply:
```
duftler@cloudshell:~$ gsutil mb -p rosco-oss-212 gs://mjd-test-123 \
>   && gsutil rm -r gs://mjd-test-123 \
>   && gsutil mb -p rosco-oss-212-test gs://mjd-test-123
Creating gs://mjd-test-123/...
Removing gs://mjd-test-123/...
Creating gs://mjd-test-123/...
duftler@cloudshell:~$
```